### ISSUE TITLE
Arc: @Default qualifier on Config class gets dropped

### ISSUE THREAD

**Issue by JHahnHRO** [quarkusio/quarkus#19526]
### Description
When a @ConfigProperties annotated bean is declared that also has the @Default qualifier, Arc seems to drop the @Default qualifier resulting in UnsatisfiedResolutionException

### How to reproduce

- define a bean aggregating config properties with `@ConfigProperties` AND `@Default` qualifier
```
@Dependent
@Default
@ConfigProperties(prefix = "my.companys.prefix")
public class AllMyConfig {

    @ConfigProperty(name = "greeting")
    String greeting;

    public String getGreeting() {
        return greeting;
    }
}
```
- inject that bean somewhere without any qualifiers
```
@Path("/hello")
public class ExampleResource {

    @Inject
    AllMyConfig allMyConfig;

    @GET
    @Produces(MediaType.TEXT_PLAIN)
    public String hello() {
        return allMyConfig.getGreeting();
    }
}
```

### Expected behaviour

Injection Point validation should succeed, since there is a bean with the correct type and qualifier @Default.

### Actual behaviour

It fails with an UnsatisfiedResolutionException instead

```
Caused by: javax.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type com.example.AllMyConfig and qualifiers [@Default]
	- java member: com.example.ExampleResource#allMyConfig
	- declared on CLASS bean [types=[com.example.ExampleResource, java.lang.Object], qualifiers=[@Default, @Any], target=com.example.ExampleResource]
	The following beans match by type, but none have matching qualifiers:
		- Bean [class=com.example.AllMyConfig, qualifiers=[@org.eclipse.microprofile.config.inject.ConfigProperties(prefix = "my.companys.prefix"), @Any]]
```

The exception message seems to indicate that Arc somehow drops the `@Default` qualifier from the bean.

**Comment by manovotn** [comment]
@JHahnHRO Quickly looking at what's happening under the hood, it looks like Quarkus is automatically adding a qualifier for each config and turning it into a bean. The code is here - https://github.com/quarkusio/quarkus/blob/main/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ConfigBuildStep.java#L265-L269

This makes sense because from CDI perspective you need to differentiate it somehow else some injection points might cause ambiguity. And so long as you have a qualifier on a bean, `@Default` should not be present (per spec this is only added to beans with no other qualifier).

In other words, the `@Default` you added is completely ignored and so is the `@Dependent` annotation. The class is not scanned for these annotations and is instead automagically converted into a synthetic bean with dependent scope and `@ConfigProperties` + `@Any` qualifier.

Therefore, to fix your issue, simply remove those annotations.

**Comment by JHahnHRO** [comment]
The spec also does not forbid to add the @Default qualifier. On the contrary: That is explicitly allowed and the reference implementation handles this correctly. The @Default qualifier is not the absence of qualifiers, it is a valid qualifier in and of itself. The only special property the spec requires is that it is added automatically to injection points and bean without other qualifiers. In particular, it is not prohibited to explicitly add it and it can be combined with other qualifiers on a bean.
I don't know what you mean by "some injection points may cause ambiguity". An injection  point without qualifier will be treated as if it had the @Default qualifier and a reference to the @Default qualified bean will be injected (whether or not that bean has additional qualifiers). An injection point with other qualifiers, e.g. @ConfigProperties, will have a reference to the the @ConfigProperties qualified bean instance injected. And that may or may not be the same bean! An injection point with both qualifiers will also get a reference to the same bean injected. All that is part of the spec. There is no ambiguity.

In other words: There is still a bug here! Quarkus should not ignore the @Default qualifier. If anything, it should automatically add that qualifier to injection points and/or beans.

**Comment by manovotn** [comment]
I never said Quarkus ignores the qualifier or that it cannot handle default qualifier. You can easily check with standard beans that this works.

The "issue", which isn't an issue but a design choice, is that Quarkus adds other than `@Default` qualifier to those config beans. It adds `@ConfigProperties` auomatically, so that you can then inject those configs as shown in https://quarkus.io/guides/config#inject-the-configuration

**Comment by JHahnHRO** [comment]
I'm not sure I understand what you're saying. Are you saying that the error lies not with the CDI implementation, but with the MicroProfile Config  implementation and/or spec, because MP Config expects every injection point to have a `@ConfigProperties` qualifier, whether or not the injectable Bean can be identified without that qualifier?
If so, then I'm not sure that's really the case. The MP Config Spec does not say explicitly that the qualifier is necessary. However, it does explicitly say that the prefix is not necessary on the injection point qualifier. When not given at the injection point, the prefix will be read from the qualifier on the bean. Nothing in there suggests to me that the combination with the `@Default` qualifier is not possible.

Or are you saying that the spec would allow all of this and the Quarkus team has made a decision to ignore the spec in this case?

**Comment by mkouba** [comment]
Well, I'm not an MP Config expert but all the examples mentioned in the spec use `@Inject @ConfigProperties`. The truth is that in "pure" CDI the example snippet should work the way you expect. However, `@Default` should not be used explicitly because the intention of the default qualifier is clear: _"If a bean does not explicitly declare a qualifier other than `@Named` or `@Any`, the bean has exactly one additional qualifier."_. In other words,  any explicit usage of `@Default` smells like a misuse or a workaround.

To sum it up: I think that the intention of MP spec authors was to always use the `@ConfigProperties` qualifier when injecting a relevant bean. @radcortez pls correct me if I'm wrong.

@JHahnHRO Anyway, what's the use case you're trying to solve with this snippet? I believe that that's the key to understand the whole problem.

**Comment by radcortez** [comment]
> To sum it up: I think that the intention of MP spec authors was to always use the `@ConfigProperties` qualifier when injecting a relevant bean. @radcortez pls correct me if I'm wrong.

Correct. For additional information, please check https://github.com/eclipse/microprofile-config/issues/629.

**Comment by JHahnHRO** [comment]
The use-case is: There is a class that collects *all* self-defined config properties (basically everything that's named "com.example.my.application.*" instead of "quarkus.*" from application.properties) in one central place. It is clear that the class `AllMyConfig` contains config properties. I had hoped that annotating that bean with `@Default` I can inject it without repeating `@ConfigProperties` everywhere.

**Comment by radcortez** [comment]
@JHahnHRO this is not on Quarkus. This is how it was defined in the MP Config spec. Please check https://github.com/eclipse/microprofile-config/issues/629.

Feel free to bring this up in the MP Config spec if you believe this should change.

**Comment by mkouba** [comment]
> The use-case is: There is a class that collects _all_ self-defined config properties (basically everything that's named "com.example.my.application._" instead of "quarkus._" from application.properties) in one central place. It is clear that the class `AllMyConfig` contains config properties. I had hoped that annotating that bean with `@Default` I can inject it without repeating `@ConfigProperties` everywhere.

Note that in Quarkus you can replace the `@Inject AllMyConfig allMyConfig` with ` @ConfigProperties AllMyConfig allMyConfig` as described in https://quarkus.io/guides/cdi-reference#qualified-injected-fields

**Comment by JHahnHRO** [comment]
@radcortez Alright, I will do that.

Thank you all for the discussion.
